File: src/plugins/java/org/ncollectd/api/Notification.java
// SPDX-License-Identifier: GPL-2.0-only
// SPDX-FileCopyrightText: Copyright (C) 2009 Hyperic, Inc.

package org.ncollectd.api;

import java.util.HashMap;
import java.util.Map;
import java.util.Date;

/**
 * Java representation of collectd/src/plugin.h:notfication_t structure.
 */
public class Notification
{
    public static final int FAILURE = 1;
    public static final int WARNING = 2;
    public static final int OKAY    = 4;

    public static String[] SEVERITY = {
        "FAILURE",
        "WARNING",
        "OKAY",
        "UNKNOWN"
    };

    private String _name;
    private long _time;
    private int _severity;
    private HashMap<String, String> _labels = new HashMap<>();
    private HashMap<String, String> _annotations = new HashMap<>();

    public Notification ()
    {
        this._severity = FAILURE;
    }

    public Notification (String name, int severity, long time,
                         HashMap<String, String> labels, HashMap<String, String> annotations)
    {
        this._name = name;
        this._severity = severity;
        this._time = time;
        this._labels = labels;
        this._annotations = annotations;
    }

    public void setName(String name)
    {
        this._name = name;
    }

    public String getName()
    {
        return this._name;
    }

    public void setSeverity (int severity)
    {
        if ((severity == FAILURE) || (severity == WARNING) || (severity == OKAY))
            this._severity = severity;
    }

    public int getSeverity()
    {
        return this._severity;
    }

    public String getSeverityString()
    {
        switch (this._severity) {
        case FAILURE:
            return SEVERITY[0];
        case WARNING:
            return SEVERITY[1];
        case OKAY:
            return SEVERITY[2];
        default:
            return SEVERITY[3];
        }
    }

    public void setTime(long time)
    {
        this._time = time;
    }

    public long getTime()
    {
        return this._time;
    }

    public void setLabels(HashMap<String, String> labels)
    {
        this._labels = labels;
    }

    public void addLabel(String name, String value)
    {
        this._labels.put(name, value);
    }

    public HashMap<String, String> getLabels()
    {
        return this._labels;
    }

    public void setAnnotations(HashMap<String, String> annotations)
    {
        this._annotations = annotations;
    }

    public void addAnnotation(String name, String value)
    {
        this._annotations.put(name, value);
    }

    public HashMap<String, String> getAnnotations()
    {
        return this._annotations;
    }

    public String toString()
    {
        StringBuffer sb = new StringBuffer();

        sb.append(_name);
        if (this._labels.isEmpty()) {
            sb.append("{}");
        } else {
            int i = 0;
            sb.append("{");
            for (Map.Entry<String, String> set : this._labels.entrySet()) {
                if (i != 0)
                    sb.append(",");
                sb.append(set.getKey()).append("=");
                sb.append("\"").append(set.getValue()).append("\"");
                i++;
            }
            sb.append("}");
        }

        if (!this._annotations.isEmpty()) {
            int i = 0;
            sb.append("{");
            for (Map.Entry<String, String> set : this._annotations.entrySet()) {
                if (i != 0)
                    sb.append(",");
                sb.append(set.getKey()).append("=");
                sb.append("\"").append(set.getValue()).append("\"");
                i++;
            }
        }

        switch (this._severity) {
        case FAILURE:
            sb.append(" FAILURE ");
        case WARNING:
            sb.append(" WARNING ");
        case OKAY:
            sb.append(" OKAY ");
        default:
            sb.append(" UNKNOWN ");
        }

        sb.append(this._time);

        return sb.toString();
    }
}

File: src/plugins/java/notification.c
// SPDX-License-Identifier: GPL-2.0-only
// SPDX-FileCopyrightText: Copyright (C) 2009  Florian octo Forster
// SPDX-FileCopyrightText: Copyright (C) 2008  Justo Alonso Achaques
// SPDX-FileContributor: Florian octo Forster <octo at collectd.org>
// SPDX-FileContributor: Justo Alonso Achaques <justo.alonso at gmail.com>

#include "plugin.h"
#include "libutils/common.h"
#include <jni.h>

#pragma GCC diagnostic ignored "-Wpedantic"

#if !defined(JNI_VERSION_1_2)
#error "Need JNI 1.2 compatible interface!"
#endif

#include "jutil.h"

extern jcached_refs_t gref;

/* Convert a notification_t to a org/ncollectd/api/Notification */
jobject ctoj_notification(JNIEnv *jvm_env, const notification_t *n)
{
    /* Create a new instance. */
    jobject o_notification = (*jvm_env)->NewObject(jvm_env, gref.notification.class,
                                                            gref.notification.constructor);
    if (o_notification == NULL) {
        PLUGIN_ERROR("Creating a new Notification instance failed.");
        return NULL;
    }

    int status = ctoj_string(jvm_env, n->name, o_notification, gref.notification.set_name);
    if (status != 0) {
        PLUGIN_ERROR("ctoj_string (setName) failed.");
        (*jvm_env)->DeleteLocalRef(jvm_env, o_notification);
        return NULL;
    }

    /* Set the `time' member. Java stores time in milliseconds. */
    (*jvm_env)->CallVoidMethod(jvm_env, o_notification, gref.notification.set_time,
                                        (jlong)CDTIME_T_TO_MS(n->time));

    /* Set the `severity' member.. */
    (*jvm_env)->CallVoidMethod(jvm_env, o_notification, gref.notification.set_severity,
                                        (jint)n->severity);

    if (n->label.num > 0 ) {
        jobject o_labels = ctoj_label_set_object(jvm_env, &n->label);
        if (o_labels == NULL) {
            PLUGIN_ERROR("ctoj_label_set_object failed.");
            (*jvm_env)->DeleteLocalRef(jvm_env, o_notification);
           return NULL;
        }
        (*jvm_env)->CallVoidMethod(jvm_env, o_notification, gref.notification.set_labels,
                                            o_labels);
        (*jvm_env)->DeleteLocalRef(jvm_env, o_labels);
    }

    if (n->annotation.num > 0 ) {
        jobject o_annotations = ctoj_label_set_object(jvm_env, &n->annotation);
        if (o_annotations == NULL) {
            PLUGIN_ERROR("ctoj_label_set_object failed.");
            (*jvm_env)->DeleteLocalRef(jvm_env, o_notification);
           return NULL;
        }
        (*jvm_env)->CallVoidMethod(jvm_env, o_notification, gref.notification.set_annotations,
                                            o_annotations);
        (*jvm_env)->DeleteLocalRef(jvm_env, o_annotations);
    }

    return o_notification;
}

/* Convert a org/ncollectd/api/Notification to a notification_t. */
int jtoc_notification(JNIEnv *jvm_env, notification_t *n, jobject object_ptr)
{
    char *name = NULL;
    int status = jtoc_string(jvm_env, &name, object_ptr, gref.notification.get_name);
    if (status != 0) {
        PLUGIN_ERROR("jtoc_string (getName) failed.");
        return -1;
    }
    n->name = name;

    jlong tmp_long = (*jvm_env)->CallLongMethod(jvm_env, object_ptr, gref.notification.get_time);
    /* Java measures time in milliseconds. */
    n->time = MS_TO_CDTIME_T(tmp_long);

    jint tmp_int = (*jvm_env)->CallIntMethod(jvm_env, object_ptr, gref.notification.get_severity);
    n->severity = (int)tmp_int;

    status = jtoc_label_set(jvm_env, &n->label, object_ptr, gref.notification.get_labels);
    if (status != 0) {
        PLUGIN_ERROR("jtoc_label_set (getLabels) failed.");
        return -1;
    }

    status = jtoc_label_set(jvm_env, &n->annotation, object_ptr, gref.notification.get_annotations);
    if (status != 0) {
        PLUGIN_ERROR("jtoc_label_set (getAnnotation) failed.");
        return -1;
    }

    return 0;
}
